internal/database/roles.go

```diff
@@ -0,0 +1,256 @@
+		sqlf.Join(roleColumns, ", "),
+		whereClause,
+	)
+
+	role, err := scanRole(r.QueryRow(ctx, q))
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, &RoleNotFoundErr{ID: opts.ID}
+		}
+		return nil, errors.Wrap(err, "scanning role")
+	}
+
+	return role, nil
+}
+
+func scanRole(sc dbutil.Scanner) (*types.Role, error) {
+	var role types.Role
+	if err := sc.Scan(
+		&role.ID,
+		&role.Name,
+		&role.ReadOnly,
+		&role.CreatedAt,
+		&dbutil.NullTime{Time: &role.DeletedAt},
+	); err != nil {
+		return nil, err
+	}
+
+	return &role, nil
+}
+
+const roleListQueryFmtstr = `
+SELECT
+	%s
+FROM roles
+WHERE %s
+`
+
+func (r *roleStore) List(ctx context.Context, opts RolesListOptions) ([]*types.Role, error) {
+	roles := make([]*types.Role, 0, 20)
+
+	scanFunc := func(rows *sql.Rows) error {
+		role, err := scanRole(rows)
+		if err != nil {
+			return err
+		}
+		roles = append(roles, role)
+		return nil
+	}
+
+	err := r.list(ctx, opts, sqlf.Join(roleColumns, ", "), scanFunc)
+	return roles, err
+}
+
+func (r *roleStore) list(ctx context.Context, opts RolesListOptions, selects *sqlf.Query, scanRole func(rows *sql.Rows) error) error {
+	var whereClause = sqlf.Sprintf("deleted_at IS NULL")
+
+	q := sqlf.Sprintf(roleListQueryFmtstr, selects, whereClause)
+
+	if opts.LimitOffset != nil {
+		q = sqlf.Sprintf("%s\n%s", q, opts.LimitOffset.SQL())
+	}
+
+	rows, err := r.Query(ctx, q)
+	if err != nil {
+		return errors.Wrap(err, "error running query")
+	}
+	defer rows.Close()
+	for rows.Next() {
+		if err := scanRole(rows); err != nil {
+			return err
+		}
+	}
+
+	return rows.Err()
+}
+
+const roleCreateQueryFmtStr = `
+INSERT INTO
+	roles (
+		name,
+		readonly
+	)
+	VALUES (
+		%s,
+		%s
+	)
+	RETURNING %s
+
+`
+
+func (r *roleStore) Create(ctx context.Context, name string, readonly bool) (_ *types.Role, err error) {
+	q := sqlf.Sprintf(
+		roleCreateQueryFmtStr,
+		name,
+		readonly,
+		// Returning
+		sqlf.Join(roleColumns, ", "),
+	)
+
+	role, err := scanRole(r.QueryRow(ctx, q))
+	if err != nil {
+		return nil, errors.Wrap(err, "scanning role")
+	}
+
+	return role, nil
+}
+
+func (r *roleStore) Count(ctx context.Context, opts RolesListOptions) (c int, err error) {
+	opts.LimitOffset = nil
+	err = r.list(ctx, opts, sqlf.Sprintf("COUNT(1)"), func(rows *sql.Rows) error {
+		return rows.Scan(&c)
+	})
+	return c, err
+}
+
+const roleUpdateQueryFmtstr = `
+UPDATE roles
+SET
+    name = %s
+WHERE
+	id = %s
+RETURNING
+	%s
+`
+
+func (r *roleStore) Update(ctx context.Context, role *types.Role) (*types.Role, error) {
+	q := sqlf.Sprintf(roleUpdateQueryFmtstr, role.Name, role.ID, sqlf.Join(roleColumns, ", "))
+
+	updated, err := scanRole(r.QueryRow(ctx, q))
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, &RoleNotFoundErr{ID: role.ID}
+		}
+		return nil, errors.Wrap(err, "scanning role")
+	}
+	return updated, nil
+}
+
+const roleDeleteQueryFmtStr = `
+UPDATE roles
+SET
+	deleted_at = NOW()
+WHERE id = %s AND readonly = false
```

Comment:
```suggestion
WHERE id = %s AND NOT readonly
```
reads a bit nicer but super nitpicky 😬